utils/tra_utils.py

```diff
@@ -0,0 +1,213 @@
+            buf.append(line[:line_comment_idx])
+            return ''.join(buf), in_comment
+
+    if block_comment_idx != -1:
+        buf.append(line[:block_comment_idx])
+        return _find_content(line[block_comment_idx+2:], True, buf=buf)
+
+    buf.append(line)
+    return ''.join(buf), in_comment
+
+
+class DFile(WeiduFile):
+    def gather_string_identifiers(self) -> Sequence[str]:
+        pat = re.compile('@[0-9]*')
+        all_idents = set()
+        in_comment = False
+        for line in self.io:
+            content, in_comment = _find_content(line, in_comment)
+            idents = pat.findall(content)
+            all_idents |= set(idents)
+        all_idents_seq = list(all_idents)
+        all_idents_seq.sort()
+        return all_idents_seq
+
+    def get_tra_file_paths(self) -> Sequence[Path]:
+        root_path = self.root_path
+        tra_dir = root_path / 'translations'
+        if not tra_dir.exists():
+            raise RuntimeError(f"Translation directory doesn't exist for file: '{self.name}' at root '{self.root_path}'")
+        if not tra_dir.is_dir():
+            raise RuntimeError(f"'translations' isn't a directory for file: '{self.name}' at root '{self.root_path}'")
+
+        lang_dirs = [child for child in tra_dir.iterdir() if child.is_dir()]
+        tra_name = self.name.replace('.d', '.tra')
+        tra_files_found = []
+        tra_files_missing = []
+        for lang_dir in lang_dirs:
+            tra_file = lang_dir / tra_name
+            if tra_file.exists():
+                tra_files_found.append(tra_file)
+            else:
+                tra_files_missing.append(tra_file)
+
+        if tra_files_missing:
+            tra_files_missing_str = ''.join([
+                f'\n * {fname}' for fname in tra_files_missing
+            ])
+            raise RuntimeError(f"Missing translation files:{tra_files_missing_str}")
+
+        return tra_files_found
+
+    def check_string_identifiers(self) -> None:
+        idents = set(self.gather_string_identifiers())
+
+        # if the file contains no string identifiers,
+        # it doesn't require .tra files, as they would be empty anyway
+        if not idents:
+            return
+
+        tra_files = self.get_tra_file_paths()
+        for tra_file_path in tra_files:
+            tra_dir = tra_file_path.parent.name
+            with TraFile.from_path(tra_file_path) as tra_file:
+                ident_strings = tra_file.gather_strings()
+            tra_idents = set(ident_strings.keys())
+            missing_idents = idents - tra_idents
+            if missing_idents:
+                missing_files_str = ''.join([
+                    f'\n * {ident}' for ident in missing_idents
+                ])
+                raise RuntimeError(f"Missing identifiers for file '{self.name}' in '{tra_dir}/{tra_file_path.name}': {missing_files_str}")
+
+
+def _get_whole_string(io: TextIO, line: str) -> str:
+    string_parts = []
+    where_end = line.find('~')
```

Comment:
i'd move this identifier `~` to some separate variable/const

Reply:
This piece was removed, as a multi-line regex now handles it.